File: app/services/ai/quotes/engine.py
from __future__ import annotations

"""Unified insurance quote engine."""

from app.models.ai_schemas import InsuranceType, QuoteRequest, QuoteResult
from app.services.ai.quotes.auto import calculate_auto_quote
from app.services.ai.quotes.home import calculate_home_quote
from app.services.ai.quotes.life import calculate_life_quote

_CALCULATORS = {
    InsuranceType.AUTO: calculate_auto_quote,
    InsuranceType.HOME: calculate_home_quote,
    InsuranceType.LIFE: calculate_life_quote,
}


def generate_quote(request: QuoteRequest) -> QuoteResult:
    """Route a quote request to the appropriate calculator."""
    calculator = _CALCULATORS.get(request.insurance_type)
    if calculator is None:
        raise ValueError(f"Unsupported insurance type: {request.insurance_type}")
    return calculator(request)


def generate_bundle_quotes(requests: list[QuoteRequest]) -> list[QuoteResult]:
    """Generate quotes for multiple lines with bundle discount."""
    from app.services.ai.quotes.rate_tables import MULTI_LINE_DISCOUNT

    results = [generate_quote(r) for r in requests]

    if len(results) > 1:
        discount_rate = MULTI_LINE_DISCOUNT * (len(results) - 1)
        discount_rate = min(discount_rate, 0.25)  # cap at 25%

        for result in results:
            original = result.annual_premium
            result.annual_premium = round(original * (1 - discount_rate), 2)
            result.monthly_premium = round(result.annual_premium / 12, 2)
            result.breakdown.discounts["multi_line_bundle"] = round(original * discount_rate, 2)
            result.breakdown.total_premium = result.annual_premium

    return results

File: frontend/src/lib/quotes/engine.ts
/**
 * Quote engine — routes to insurance-type-specific calculators.
 * Supports individual and bundle quotes.
 */

import { calculateAutoQuote, type AutoQuoteInput } from "./auto";
import { calculateHomeQuote, type HomeQuoteInput } from "./home";
import { calculateLifeQuote, type LifeQuoteInput } from "./life";
import { MULTI_LINE_DISCOUNT, MAX_BUNDLE_DISCOUNT } from "./rate-tables";
import type { QuoteResult, BundleQuoteResult } from "./types";

export type QuoteInput =
  | { type: "auto"; data: AutoQuoteInput }
  | { type: "home"; data: HomeQuoteInput }
  | { type: "life"; data: LifeQuoteInput };

/** Generate a single insurance quote. */
export function generateQuote(input: QuoteInput): QuoteResult {
  switch (input.type) {
    case "auto":
      return calculateAutoQuote(input.data);
    case "home":
      return calculateHomeQuote(input.data);
    case "life":
      return calculateLifeQuote(input.data);
    default:
      throw new Error(`Unsupported insurance type: ${(input as { type: string }).type}`);
  }
}

/** Generate bundle quotes with multi-line discount. */
export function generateBundleQuotes(inputs: QuoteInput[]): BundleQuoteResult {
  if (inputs.length === 0) {
    throw new Error("At least one quote input is required");
  }

  const quotes = inputs.map(generateQuote);
  const totalBeforeDiscount = quotes.reduce((sum, q) => sum + q.annualPremium, 0);

  const lineCount = Math.min(quotes.length, 5);
  const discountRate = Math.min(
    MULTI_LINE_DISCOUNT[lineCount] ?? 0,
    MAX_BUNDLE_DISCOUNT
  );

  const savings = Math.round(totalBeforeDiscount * discountRate);
  const totalAfterDiscount = totalBeforeDiscount - savings;
  const monthlyAfterDiscount = Math.round((totalAfterDiscount / 12) * 100) / 100;

  return {
    quotes,
    bundleDiscount: discountRate,
    totalAnnualBeforeDiscount: totalBeforeDiscount,
    totalAnnualAfterDiscount: totalAfterDiscount,
    totalMonthlyAfterDiscount: monthlyAfterDiscount,
    savings,
  };
}

// Re-export types and calculators
export type { QuoteResult, BundleQuoteResult, PremiumBreakdown } from "./types";
export type { AutoQuoteInput } from "./auto";
export type { HomeQuoteInput } from "./home";
export type { LifeQuoteInput } from "./life";
